### PR TITLE
Add FLAC export and related options

### PR DESCRIPTION
**Pull request by irrenhaus3**
Implements #3679 by adding a FLAC rendering device to LMMS' Project renderer and related GUI configuration widgets in the export window.  
The renderer is implemented in terms of libsndfile and **hardens the dependency on it**, requiring a minimal version of 1.0.18.  
Additionally, since the FLAC compression level settings was not supported in libsndfile until version 1.0.26, a new CPP flag is added, called LMMS_HAVE_SF_COMPLEVEL. It is undefined if the sndfile version is less than 1.0.26. The "Compression Level" GUI element is only visible (and the setting received from it is only respected in the renderer) if LMMS_HAVE_SF_COMPLEVEL is defined.  
  
Unrelated to this new feature, this PR also introduces a minor bug fix regarding the output file extension. If the output format doesn't match the file extension set by the user before the export dialog window opens, the extension will automatically be adjusted when the rendering process starts.

### CODE REVIEW

File: src/core/audio/AudioFileFlac.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	}
+
+#ifdef LMMS_HAVE_SF_COMPLEVEL
+	double compression = getOutputSettings().getCompressionLevel();
+	sf_command(m_sf,SFC_SET_COMPRESSION_LEVEL,&compression,sizeof(double));
+#endif
+
+	m_sf = sf_open(
+#ifdef LMMS_BUILD_WIN32
+		outputFile().toLocal8bit().constData(),
```
Comment (zonkmachine) [inline src/core/audio/AudioFileFlac.cpp]: `toLocal8bit()` -> `toLocal8Bit()`

Comment (irrenhaus3) [inline src/core/audio/AudioFileFlac.cpp]: Lol, and here I was wondering why travis exploded. Thanks!

**Comment by zonkmachine** [comment]
> #include <QtGui/QAction>

For this you should be able to just go:
`#include <QAction>`

**Comment by zonkmachine** [comment]
> plugins/sf2_player/ui_patches_dialog.h. The offending line is
> #include <QtGui/QAction>

I actually can't find that file or the #include line anywhere.

File: src/gui/ExportProjectDialog.cpp [hunk 2]
```diff
@@ -79,30 +79,37 @@ ExportProjectDialog::ExportProjectDialog( const QString & _file_name,
...
 		}
 	}
+	
+	int const MAX_LEVEL=8;
+	for(int i=0; i<=MAX_LEVEL; ++i)
+	{
+		QString info="";
+		if (i==0){ info = tr("(fastest)", nullptr, QApplication::UnicodeUTF8); }
+		if (i==4){ info = tr("(default)", nullptr, QApplication::UnicodeUTF8); }
+		if (i==MAX_LEVEL){ info = tr("(smallest)", nullptr, QApplication::UnicodeUTF8); }
```
Comment (PhysSong) [inline src/gui/ExportProjectDialog.cpp]: @irrenhaus3 I suggest using `tr("(something)")` without additional variable(s).

Comment (PhysSong) [inline src/gui/ExportProjectDialog.cpp]: Here Qt5 build fails, too.

**Comment by irrenhaus3** [comment]
My bad about that autogenerated SF2player GUI header. Turns out I didn't properly clean up my build directory and that particular header was left over from a previous QT4 build. Oops...

File: src/gui/ExportProjectDialog.cpp (hunk 2 above)
Comment (irrenhaus3) [inline src/gui/ExportProjectDialog.cpp]: Good catch, thank you. The symbol was removed in the newest commit.

File: include/AudioFileFlac.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
+/*
+ * AudioFileFlac.h - Audio device which encodes a wave stream into a FLAC file.
+ */
```
Comment (PhysSong) [inline include/AudioFileFlac.h]: You should put the copyright information here.

File: src/core/audio/AudioFileFlac.cpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
+#include <memory>
```
Comment (PhysSong) [inline src/core/audio/AudioFileFlac.cpp]: Here, too.

**Comment by PhysSong** [comment]
However, the rest things work great. I tested it with both 1.0.25 and 1.0.28.

**Comment by PhysSong** [comment]
I found one more bug just now: It breaks "Export tracks" path setting. If I choose directory '/foo/bar', it tries to export tracks to '/foo/bar.mp3/1_asdf.mp3'.

**Comment by PhysSong** [comment]
One more request: please add WANT_FLAC option to CMakeLists.txt and let the code be built with libsndfile 1.0.11~1.0.17 without flac export.

**Comment by tresf** [comment]
> let the code be built with libsndfile 1.0.11~1.0.17 without flac export.

As stated in the PR, `sndfile >= 1.0.18` will be the new minimum version on our master branch.  This was discussed on Discord.  Legacy systems that do not have `sndfile >= 1.0.18` will have to use `stable-1.2` or older.  Why?  Because Ubuntu 10.04 offered 1.0.21, and that distribution is now 7 years old and was end of life 2 years ago. :)

**Comment by PhysSong** [comment]
@tresf Quite true.
@irrenhaus3 However, WANT_FLAC and status message in CMakeLists.txt is still needed.

**Comment by tresf** [comment]
> However, WANT_FLAC and status message in CMakeLists.txt is still needed.

Why?

**Comment by PhysSong** [comment]
>Why?
@tresf Both OGG and MP3 have such things, so user can enable/disable such features manually.
https://github.com/LMMS/lmms/blob/stable-1.2/CMakeLists.txt#L55

**Comment by tresf** [comment]
Because they are (or were) separate libraries.  Don't add a toggle for historical purposes please.  Instead, remove OGG.

**Comment by PhysSong** [comment]
>Because they are (or were) separate libraries.

Yes.
>Don't add a toggle for historical purposes please.

I don't agree. They were added because of historical purposes, but I think they are still useful.
I can disable some features if I don't want them, for build speed and some other purposes.
>Instead, remove OGG toggle.

I agree that it should be, if FLAC toggle is not added, for consistency.

**Comment by irrenhaus3** [comment]
@PhysSong: I agree with @tresf on this one. I really can't picture a use-case where a post-1.2 LMMS end user would want to disable FLAC, but keep WAV, especially since they're both implemented in terms of the same library. While I agree that it might make some sense in development and testing, this particular feature doesn't noticably increase build times or interfere with unrelated system components.  
As such, I'd like to refrain from implementing a WANT_FLAC mode for now. If the need for it ever arises, it can easily be done anyway. But in this PR, I'll only do it if the majority of developers agrees on it.

**Comment by PhysSong** [comment]
@irrenhaus3 Then could you wait for other developers' opinion for CMakeLists.txt issue?
However, I think the code needs no more change(s) after @zonkmachine's [suggestions](https://github.com/LMMS/lmms/pull/3731#issuecomment-318950581) are applied.

**Comment by zonkmachine** [comment]
Removing the ogg toggle option is better than adding a flac toggle. Keep it simple.

File: src/core/ProjectRenderer.cpp [hunk 5]
```diff
@@ -42,6 +43,11 @@ const ProjectRenderer::FileEncodeDevice ProjectRenderer::fileEncodeDevices[] =
 	{ ProjectRenderer::WaveFile,
 		QT_TRANSLATE_NOOP( "ProjectRenderer", "WAV-File (*.wav)" ),
 					".wav", &AudioFileWave::getInst },
+    { ProjectRenderer::FlacFile,
```
Comment (tresf) [inline src/core/ProjectRenderer.cpp]: @irrenhaus3 this needs formatting, spaces versus tabs. :)

**Comment by tresf** [comment]
A follow-up task to merging this should be to update the required `libsndfile` version on our [`Compiling`](https://github.com/LMMS/lmms/wiki/Compiling#libraries) wiki page.

**Comment by PhysSong** [comment]
@irrenhaus3 Then could you remove WANT_OGG toggle and related flags?

**Comment by irrenhaus3** [comment]
Update for the interested reader not on the LMMS Discord server:  
In a discussion on the developers channel, we found that refactoring the OGG exporter to use sndfile and removing the WANT_OGGVORBIS flag are separate issues from FLAC export and will be handled in their own PR.

File: src/core/audio/AudioFileFlac.cpp [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+#ifdef LMMS_BUILD_WIN32
+		outputFile().toLocal8Bit().constData(),
+#else
+		outputFile().toUtf8().constData(),
+#endif
+		SFM_WRITE,
+		&m_sfinfo
+	);
+
+	sf_command(m_sf,SFC_SET_CLIPPING,nullptr,SF_TRUE);
```
Comment (tresf) [inline src/core/audio/AudioFileFlac.cpp]: It is conventional to put spaces between parameters.

Comment (PhysSong) [inline src/core/audio/AudioFileFlac.cpp]: Like `foo(b, a, r);`.

**Comment by PhysSong** [comment]
I recommend adding a status message for FLAC, like WAV one. It's not necessary, but I think having this will be better.

**Comment by gi0e5b06** [comment]
+1 for WANT_FLAC
Anything that use external libraries should be flagged.

**Comment by PhysSong** [comment]
Great! Now everything looks fine.
I'll merge it soon, if there's no objection.

**Comment by zthurman** [comment]
I'm probably being a dumb but I'm not seeing this in the front end options on lmms 1.2.0-rc7 running on Arch or the lmms-git version 1.2.0-rc7.11.

Am I understanding correctly from the above that I have to build from source and tweak out a bit in order to use this enhancement?

**Comment by zonkmachine** [comment]
> I'm probably being a dumb but I'm not seeing this in the front end options on lmms 1.2.0-rc7 running on Arch or the lmms-git version 1.2.0-rc7.11.

It's because this has been added to master and will be in lmms-1.3.

> Am I understanding correctly from the above that I have to build from source and tweak out a bit in order to use this enhancement?

You either build `master` or you can cherry-pick commit https://github.com/LMMS/lmms/commit/e9a406311969397a9d4219eb80671d27db826e87 to stable-1.2 .
I've seen unofficial builds around for master that you could try out but I don't know precisely where or if they are up to date. Check out our support channel at discord, https://discord.gg/3sc5su7 (`#support`), and ask there. You'll get help much sooner there I think. This is a closed pull request and late comments on those tend to get little attention.

**Pull request merged**